### DISCUSSION TITLE
Access to the underlying tungstenite WebSocketStream config

### DISCUSSION THREAD

**Discussion by mr-ns** [poem-web/poem#890] [Q&A] [answered]
I would like to be able to configure things like the max message size when using Poem's WebSocket. This is available on the [underlying tungstenite WebSocketStream that Poem wraps](https://docs.rs/poem/3.1.0/src/poem/web/websocket/stream.rs.html#14-16), but I do not see it exposed through Poem. The end result is something like the following, which seems … less than ideal:
```rust
#[handler]
pub async fn handler(ws: WebSocket) -> impl IntoResponse {
    ws.on_upgrade(|mut socket| async move {
        unsafe {
            let poem_stream: *const poem::web::websocket::WebSocketStream = &socket;
            let tung_stream = poem_stream as *const tokio_tungstenite::WebSocketStream<poem::Upgraded>;

            let config_const: *const tungstenite::protocol::WebSocketConfig = (*tung_stream).get_config();
            let config = config_const as *mut tungstenite::protocol::WebSocketConfig;

            (*config).max_message_size = Some(32 << 20);
            (*config).max_frame_size = Some(8 << 20);
        }
...
```
The pointers do not bother me (hey, I grew up on C), but two other things do:
1. The `let tung_stream = poem_stream as ...` cast assumes the tungstenite WebSocketStream is the first member in the Poem WebSocketStream; that is fragile at best.
2. Coercing mutability onto the reference returned by `get_config()` could have unintended side effects. After all, if the tungstenite folks wanted the stream config to be changeable they would have returned a mutable reference in the first place.

Have I overlooked some "blessed" way to configure the WebSocketStream?

**Comment by sunli829** [comment]
Added `WebSocket::config` method in `poem v3.1.1`

**Comment by mr-ns** [comment, marked as answer]
Cool! So we would call `WebSocket::config()` in the handler right before `WebSocket::on_upgrade()` ?

**Reply by sunli829** [reply]
yes

**Reply by mr-ns** [reply]
Sweet! Thank you!

**Comment by mr-ns** [comment]
@sunli829 — Thank you again for your quick attention to this request. Now that I am using this I have come across another desire; that is, the ability to _get_ the current configuration. For reference, the tungstenite library has a `get_config()` method on both the [WebSocket](https://docs.rs/tungstenite/latest/tungstenite/protocol/struct.WebSocket.html#method.get_config) and [WebSocketStream](https://docs.rs/tokio-tungstenite/latest/tokio_tungstenite/struct.WebSocketStream.html#method.get_config). Can we expose those through the Poem layer? In particular, the ability to get the config from the stream would be very nice.

Thank you!

**Reply by mr-ns** [reply]
@sunli829 — Any thoughts here? Should I open a new request for this?

**Reply by mr-ns** [reply]
@sunli829 — I implemented this for `WebSocketStream` and submitted PR #900. The implementation (intentionally) mirrors that of the [tungstenite equivalent](https://docs.rs/tokio-tungstenite/latest/tokio_tungstenite/struct.WebSocketStream.html#method.get_config).
